Gradcheck for complex

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33507 Gradcheck for complex**
* #33361 torch.tensor can infer complex dtype now
* #33268 Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float

Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

anjali411 committed:
gradcheck

[ghstack-poisoned]

anjali411 committed:
Update on "Gradcheck for complex"


Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

[ghstack-poisoned]

anjali411 committed:
Update on "Gradcheck for complex"


Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -49,6 +49,11 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+
+    is_complex = input[0].dtype.is_complex
+    if is_complex:
+        eps = eps * 1j
```
I think you want `eps + eps * 1j` here to match the previous discussions we had.

anjali411 committed:
Update on "Gradcheck for complex"


Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

[ghstack-poisoned]

anjali411 committed:
Update on "Gradcheck for complex"


Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

[ghstack-poisoned]

anjali411 committed:
Update on "Gradcheck for complex"


Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

[ghstack-poisoned]

anjali411 committed:
Update on "Gradcheck for complex"


Differential Revision: [D19975615](https://our.internmc.facebook.com/intern/diff/D19975615)

[ghstack-poisoned]

anjali411 commented:
landed https://github.com/pytorch/pytorch/pull/35238
